### PR TITLE
Fix chain of trust certification

### PR DESCRIPTION
**Pull request by mhpnguyen**
Apply special logic to handle chain of trust.  It will validate the passing cert against trust store to make sure  all the chain exist.
Optional: @trantang @ptambellini01

### CODE REVIEW

File: Product/Production/Common/CONNECTCoreLib/src/test/java/gov/hhs/fha/nhinc/callback/cxf/CertificateChainValidatorTest.java [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+    }
+    @Test
+    public final void testValidateCert(){
+        X509Certificate certToCheck = getTestCert("/epic.cer");
+        Assert.assertTrue("Our truststore should contain this cert",certValidator.validateCert(certToCheck));
+    }
+    @Test
+    public final void testValidRoot(){
+        /*certValidator = null;
+        certManager = null;
```
Comment (TabassumJafri) [inline Product/Production/Common/CONNECTCoreLib/src/test/java/gov/hhs/fha/nhinc/callback/cxf/CertificateChainValidatorTest.java]: remove commented out code

File: Product/Production/Common/CONNECTCoreLib/src/main/java/gov/hhs/fha/nhinc/callback/cxf/CertificateChainValidator.java [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+     */
+    public boolean validateCert(X509Certificate certToCheck) {
+        String leafIssuerString = certToCheck.getIssuerX500Principal().getName();
+        LOG.debug("Issue Name for validated Cert {}", leafIssuerString);
+        // find all intermediate in truststore if exist
+        List<X509Certificate> allCerts = getAllCertByIssuer(certificateManager.getTrustStore(), leafIssuerString);
+        // check if cert is chain of trust
+        if (isPartChainOfTrust(certToCheck)) {
+            LOG.debug("Cert serial number to verify against our trustStore {}",
+                certToCheck.getSerialNumber().toString(16));
```
Comment (TabassumJafri) [inline Product/Production/Common/CONNECTCoreLib/src/main/java/gov/hhs/fha/nhinc/callback/cxf/CertificateChainValidator.java]: You can use BASE16 constant .. using just 16 is very vague.

File: Product/Production/Common/CONNECTCoreLib/src/test/java/gov/hhs/fha/nhinc/callback/cxf/CertificateChainValidatorTest.java [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+    @Test
+    public final void testInvalidCert(){
+        X509Certificate certToCheck = getTestCert("/HOST1.cer");
+        Assert.assertFalse("Our truststore should not contain",certValidator.validateCert(certToCheck));
+    }
+    @Test
+    public final void testValidOnlyInterAndRootCert(){
+        //This test cert that only has intermediate and root.
+        X509Certificate certToCheck = getTestCert("/entrust_intermediate_old.cer");
+        Assert.assertTrue("Our truststore should this chain of trust",certValidator.validateCert(certToCheck));
```
Comment (TabassumJafri) [inline Product/Production/Common/CONNECTCoreLib/src/test/java/gov/hhs/fha/nhinc/callback/cxf/CertificateChainValidatorTest.java]: Need more meaningful message

File: Product/Production/Common/CONNECTCoreLib/src/main/java/gov/hhs/fha/nhinc/callback/cxf/CONNECTSamlAssertionValidator.java [hunk 4]
```diff
@@ -304,11 +307,40 @@ protected void checkSignedAssertion(AssertionWrapper assertion, RequestData data
...
+                    for (X509Certificate cert: certList){
+                        String serialNumber = cert.getSerialNumber().toString(16);
+                        if (chainValidator.validateCert(cert)){
+                            LOG.info("A certs serial number in saml already in the truststore {}",serialNumber);
+                        }else{
+                            LOG.error("Unable to find cert serial number in the truststore {}",serialNumber);
+                            throw e;
+                        }
+                    }
+
```
Comment (TabassumJafri) [inline Product/Production/Common/CONNECTCoreLib/src/main/java/gov/hhs/fha/nhinc/callback/cxf/CONNECTSamlAssertionValidator.java]: Remove extra line spaces

File: Product/Production/Common/CONNECTCoreLib/src/main/java/gov/hhs/fha/nhinc/callback/cxf/CertificateChainValidator.java [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+            for (X509Certificate cert : allCerts) {
+                if (cert.equals(certToCheck)) {
+                    LOG.debug("Found the cert in trust store with serial number {}", cert.getSerialNumber()
+                        .toString(16));
+                    return true;
+                }
+            }
+            return false;
+        }
+        return false;// by default
```
Comment (TabassumJafri) [inline Product/Production/Common/CONNECTCoreLib/src/main/java/gov/hhs/fha/nhinc/callback/cxf/CertificateChainValidator.java]: remove the comments here .. add them as method level javadoc

File: Product/Production/Common/Properties/src/main/resources/gateway.properties [hunk 6]
```diff
@@ -4,6 +4,8 @@ localHomeCommunityId=urn:oid:1.1
 #Device Id of local gateway, used for Patient Discovery
 localDeviceId=1.1
 
+# True if handle chain of trust by comparison between Authority Key Identifier(Leaf) and Subject Key Identifier(parent)
+ValiateChainOfTrust=false
```
Comment (TabassumJafri) [inline Product/Production/Common/Properties/src/main/resources/gateway.properties]: Misspelled property

**Review by TabassumJafri — CHANGES_REQUESTED** [review]
Please see my inline comments

File: Product/Production/Common/Properties/src/main/resources/gateway.properties (hunk 6 above)
Comment (jsmith2) [reply Product/Production/Common/Properties/src/main/resources/gateway.properties]: Also misspelled validate.

File: Product/Production/Common/CONNECTCoreLib/src/main/java/gov/hhs/fha/nhinc/callback/cxf/CONNECTSamlAssertionValidator.java [hunk 7]
```diff
@@ -65,7 +68,7 @@
 public class CONNECTSamlAssertionValidator extends SamlAssertionValidator {
 
     /** The Constant LOG. */
-    private static final Logger LOG = Logger.getLogger(CONNECTSamlAssertionValidator.class);
+    private static final Logger LOG = LoggerFactory.getLogger(CONNECTSamlAssertionValidator.class);
```
Comment (jsmith2) [inline Product/Production/Common/CONNECTCoreLib/src/main/java/gov/hhs/fha/nhinc/callback/cxf/CONNECTSamlAssertionValidator.java]: Change this back to to log4j.  I don't think the SLF4J change was made during 4.5.

File: Product/Production/Common/CONNECTCoreLib/src/main/java/gov/hhs/fha/nhinc/callback/cxf/CONNECTSamlAssertionValidator.java [hunk 8]
```diff
@@ -304,11 +307,40 @@ protected void checkSignedAssertion(AssertionWrapper assertion, RequestData data
...
+                    throw e;
+                }
+
             }
         }
     }
 
+    protected boolean isValidateChainOfTrust() {
+        try {
+            return propertyAccessor.getPropertyBoolean(NhincConstants.GATEWAY_PROPERTY_FILE, "ValiateChainOfTrust");
```
Comment (jsmith2) [inline Product/Production/Common/CONNECTCoreLib/src/main/java/gov/hhs/fha/nhinc/callback/cxf/CONNECTSamlAssertionValidator.java]: Validate spelled wrong. "ValidateChainOfTrust"

**Review by jsmith2 — CHANGES_REQUESTED** [review]

**Review by TabassumJafri — APPROVED** [review]
Endorsing for merge

**Review by jsmith2 — APPROVED** [review]

**Pull request merged**